metricflow/model/semantics/join_validator.py

```diff
@@ -58,34 +75,140 @@ class DataSourceJoinValidator:
+                        known_data_source_joins=data_source_joins,
+                        identifier_reference=secondary_identifier_reference,
+                    )
+                    for (
+                        secondary_joinable_data_source_reference,
+                        secondary_join_type,
+                    ) in secondary_joinable_data_source_references:
+                        if (
+                            secondary_joinable_data_source_reference.data_source_name
+                            == left_data_source_reference.data_source_name
+                        ):
+                            continue
+                        data_source_joins[secondary_joinable_data_source_reference.data_source_name] = DataSourceJoin(
+                            left_data_source_reference=left_data_source_reference,
+                            right_data_source_reference=secondary_joinable_data_source_reference,
+                            identifier_joins=[
+                                DataSourceIdentifierJoin(
+                                    identifier_reference=identifier_reference, join_type=join_type
+                                ),
+                                DataSourceIdentifierJoin(
+                                    identifier_reference=secondary_identifier_reference, join_type=secondary_join_type
+                                ),
+                            ],
+                        )
+
+        return data_source_joins
+
+    def _get_new_joinable_data_sources_for_identifier(
+        self,
+        left_data_source_reference: DataSourceReference,
+        known_data_source_joins: Dict[str, DataSourceJoin],
+        identifier_reference: IdentifierReference,
+    ) -> List[Tuple[DataSourceReference, DataSourceIdentifierJoinType]]:
+        new_joinable_data_sources: List[Tuple[DataSourceReference, DataSourceIdentifierJoinType]] = []
+        identifier_data_sources = self._data_source_semantics.get_data_sources_for_identifier(
+            identifier_reference=identifier_reference
+        )
+        for identifier_data_source in identifier_data_sources:
+            if (
+                identifier_data_source.name == left_data_source_reference.data_source_name
+                or identifier_data_source.name in known_data_source_joins
+            ):
+                continue
+
+            right_data_source_reference = DataSourceReference(data_source_name=identifier_data_source.name)
+            valid_join_type = self.get_valid_data_source_identifier_join_type(
+                left_data_source_reference=left_data_source_reference,
+                right_data_source_reference=right_data_source_reference,
+                on_identifier_reference=identifier_reference,
+            )
+            if valid_join_type is not None:
+                new_joinable_data_sources.append((right_data_source_reference, valid_join_type))
```

Comment:
If we don't need to pass this join type info from down here out to the return value the whole function becomes a pair of filters - memoization and validity.

Reply:
We do need the join type for the model page!